lib/scala/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerProcess.scala

```diff
@@ -0,0 +1,327 @@
+    context.stop(self)
+  }
+
+  /** In booting stage, the actor is retrying to connect to the server to verify
+    * that it has finished initialization.
+    *
+    * Once initialization is confirmed, it notifies the parent and proceeds to
+    * runningStage.
+    *
+    * Before initialization is confirmed the actor can also:
+    * - trigger the timeout or receive a graceful stop message which will ask
+    *   the child process to terminate gracefully,
+    * - get a notification that the child process has terminated spuriously
+    *   (possibly a crash).
+    */
+  private def bootingStage(
+    process: Process,
+    bootTimeout: Cancellable
+  ): Receive =
+    handleBootResponse(process, bootTimeout).orElse(runningStage(process))
+
+  case object AskServerIfStarted
+  case object TimedOut
+  private val retryDelay = 100.milliseconds
+
+  private def handleBootResponse(
+    process: Process,
+    bootTimeout: Cancellable
+  ): Receive = {
+    case AskServerIfStarted =>
+      val socket = Socket(descriptor.networkConfig.interface, rpcPort)
+      context.actorOf(
+        HeartbeatSession.initialProps(
+          socket,
+          retryDelay,
+          new AkkaBasedWebSocketConnectionFactory()(context.system),
+          context.system.scheduler
+        ),
+        s"initial-heartbeat-${UUID.randomUUID()}"
+      )
+    case TimedOut =>
+      self ! Stop
+    case HeartbeatReceived =>
+      context.parent ! LanguageServerProcess.ServerConfirmedFinishedBooting
+      bootTimeout.cancel()
+      context.become(runningStage(process))
+    case ServerUnresponsive =>
+      import context.dispatcher
+      context.system.scheduler.scheduleOnce(
+        retryDelay,
+        self,
+        AskServerIfStarted
+      )
+      context.become(bootingStage(process, bootTimeout))
+  }
+
+  /** When the process is running, the actor is monitoring it and managing its
+    * lifetime.
+    *
+    * If termination is requested, the process will be asked to gracefully
+    * terminate or be killed. If the process terminates (either on request or
+    * spuriously, e.g. by a crash), the parent is notified and the current actor
+    * is stopped.
+    */
+  private def runningStage(process: Process): Receive = {
+    case Stop =>
+      requestGracefulTermination(process)
+    case Kill => process.destroyForcibly()
+    case ServerTerminated(exitCode) =>
+      context.parent ! ServerTerminated(exitCode)
+      context.stop(self)
+    case ServerThreadFailed(error) => handleFatalError(error)
+  }
+
+  /** Runs the child process, ensuring that the proper locks are kept for the
+    * used engine for the whole lifetime of that process.
+    *
+    * Returns the exit code of the process. This function is blocking so it
+    * should be run in a backgroung thread.
+    */
+  private def runServer(processStarted: Promise[Process]): Int = {
+    val versionManager = makeRuntimeVersionManager(progressTracker)
```

Comment:
Please extract the whole logic responsible for executing LS processes to a separate object let's say `LanguageServerExecutor`. This component should encapsulate all details regarding spawning new processes (process API, thread pool).

Reply:
The actor should be only responsible for orchestrating the lifecycle of LS and reacting to LS events.